How many late days are you using for this assignment? 
0

In [77]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np


# Part 1: Load the dataset

In [78]:
# Load the given datasets
dataset_cat = pd.read_csv('/Users/<redacted>/Documents/GitHub/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_categorical.csv')
dataset_num = pd.read_csv('/Users/<redacted>/Documents/GitHub/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_numerical.csv')

# Print the data
# .head() displays first 5 rows of data set
display(dataset_cat.head())
display(dataset_num.head())

# displays the data set info
# dataset_num.info()
# dataset_cat.info()

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The biggest between the two different data sets are in the name, the categorical one has non number values and the numerical one has numerical values. The categorical data set is used to define things such as red blood cells, sugar, pus cells, etc. and their corresponding levels. The numerical data set shows more quantifiable data such as age, blood pressure, potassium etc. Something they could share in common would be the unique id, if they both match to one person. 

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [79]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_rows = dataset_num.duplicated()
print(duplicate_rows.sum())

# Drop these duplicate rows
numeric = dataset_num.drop_duplicates()
print(numeric.head())

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_rows = dataset_cat.duplicated()
print(duplicate_rows.sum())

categoric = dataset_cat.drop_duplicates()
print(numeric.head())

30
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  8200.0   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN     NaN   

   rbcc  
0   6.4  
1   5.5  
2   5.7  
3   4.5  
4   NaN  
30
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  8200.0   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [80]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
dataset = pd.merge(numeric, categoric, on='unique_id')
#Print the combined dataset
display(dataset)


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [81]:
# Calculate the percentage of rows that contain atleast one missing value
# dataset.isna().sum() number of missing datas from columns

missing_rows = dataset.isna().any(axis=1).sum()
# Print %

total_rows = dataset.shape[0]
percentage = (missing_rows / total_rows) * 100
print(percentage)
# Drop these rows from the dataset
dataset_cleaned = dataset.dropna()
# Print the Dataset
display(dataset_cleaned)

60.76923076923077


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
13,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
379,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
381,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [82]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
dataset_sorted = dataset_cleaned.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
display(dataset_sorted)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [83]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dataset_encoded = pd.get_dummies(dataset_sorted, columns=categorical_columns)
# Print the dataset
display(dataset_encoded)
# print(dataset_encoded.columns)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,True,False,False,True,True,False,False,True,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,True,False,True,False,True,False,True,False,False,True


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [84]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

def remove_outliers(df, columns):
    for column in columns:
        mean = df[column].mean()
        std = df[column].std()
        df = df[(df[column] > mean - 3 * std) & (df[column] < mean + 3 * std)]
    return df
# Print the dataset
dataset = remove_outliers(dataset_encoded, numerical_columns) # outliers removed
display(dataset)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,False,True,False,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,True,False,True,False,True,False,True,False,True,False
6,546225,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,...,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,True,False,True,False,True,False,True,False,False,True


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [85]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import MinMaxScaler, StandardScaler, RobustScaler

numerical_cols = dataset.select_dtypes(include=['float64', 'int64']).columns.tolist()
print("Numerical columns:", numerical_cols)

# Step 4: Choose a normalization method
# Example: Min-Max Normalization
scaler = MinMaxScaler()

# Step 5: Apply normalization
dataset[numerical_cols] = scaler.fit_transform(dataset[numerical_cols])

# Check the results
print("Normalized Data:")
display(dataset.head())
# Print the dataset

Numerical columns: ['unique_id', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc', 'al', 'su']
Normalized Data:


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.411888,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,...,True,False,True,False,True,False,False,True,True,False
1,0.031673,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,...,False,True,False,True,True,False,True,False,True,False
3,0.264544,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,...,False,True,False,True,False,True,False,True,True,False
4,0.477403,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,...,True,False,True,False,True,False,True,False,True,False
6,0.492853,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,...,True,False,True,False,True,False,True,False,True,False


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Unique_id, because it is random and does not contribute to the modeling andprediction since it is completely random and has no correlation.

In [86]:
#Remove that column
unnecessary_columns = ['unique_id', 'cad_no','appet_poor', 'pe_no', 'dm_no', 'ba_notpresent', 'pcc_notpresent', 'rbc_normal', 'pc_normal', 'ane_no', 'htn_no', 'Target_notckd']
dataset = dataset.drop(columns=unnecessary_columns)

display(dataset.head())
# Print the dataset

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,pc_abnormal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_good,pe_yes,ane_yes,Target_ckd
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,...,True,True,True,False,False,False,True,False,True,True
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,...,True,True,False,True,True,True,False,False,False,True
3,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,...,True,False,False,True,True,True,False,True,True,True
4,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,...,False,False,False,False,False,False,True,False,False,True
6,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,...,False,False,False,True,True,False,True,False,False,True


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [72]:
# Export the dataframe to a new csv file
dataset.to_csv('/Users/<redacted>/Documents/GitHub/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_cleaned_final.csv', index=False)
# Export the dataframe to a new json file
dataset.to_json('/Users/<redacted>/Documents/GitHub/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_cleaned_final.json', orient='records')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

In [73]:
top_15_rows = dataset.head(15)
display(top_15_rows)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,pc_abnormal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_good,pe_yes,ane_yes,Target_ckd
0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,...,True,True,True,False,False,False,True,False,True,True
1,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.148515,0.225806,0.217742,...,True,True,False,True,True,True,False,False,False,True
3,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,...,True,False,False,True,True,True,False,True,True,True
4,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,0.322581,0.500000,...,False,False,False,False,False,False,True,False,False,True
6,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,...,False,False,False,True,True,False,True,False,False,True
8,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,0.169355,...,True,False,False,True,False,False,True,False,False,True
10,0.716216,0.50,1.000000,0.163399,0.111111,0.066667,0.206897,0.267327,0.387097,0.532258,...,False,False,False,False,True,False,False,False,False,True
13,0.729730,0.00,0.935622,0.169935,0.160494,0.333333,0.034483,0.019802,0.064516,0.879032,...,True,True,False,True,False,False,False,False,True,True
17,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,0.225806,1.000000,...,True,False,True,False,False,False,False,False,False,True
19,0.878378,0.00,0.206009,0.751634,0.604938,0.533333,0.689655,0.366337,0.387097,0.879032,...,False,False,False,True,True,False,False,True,False,True


## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 
| Age | BP | BGR | BU | SC | SOD | POT | Hemo | PCV | WBCC | RBCC | AL | SU | RBC Abnormal | RBC Normal | PC Abnormal | PC Normal | PCC Not Present | PCC Present | BA Not Present | BA Present | HTN No | HTN Yes | DM No | DM Yes | CAD No | CAD Yes | Appet Good | Appet Poor | PE No | PE Yes | ANE No | ANE Yes | Target CKD | Target not CKD |
|------------|-----|--------------------------|--------------------------|-------------------------|-------------------------|------------------------|----------------------|----------------------|-----------------------|-----------------------|------|----|--------------|------------|-------------|-----------|------------------|-------------|----------------|------------|--------|---------|-------|--------|--------|---------|-------------|------------|-------|--------|--------|---------|-------------|-----------------|
| 0.202703 | 0.75| 0.15879828326180256 | 0.19607843137254904 | 0.16049382716049385 | 0.16666666666666696 | 0.2068965517241379 | 0.05940594059405946 | 0.0 | 0.653225806451613 | 0.25714285714285706 | 1.0 | 0.0| False | True | True | False | False | True | False | True | True | False | True | False | True | False | True | False | True | False | True | False | True | False |
| 0.905405 | 1.0 | 0.9656652360515021 | 0.5228758169934641 | 0.6419753086419753 | 0.666666666666667 | 0.0 | 0.14851485148514842 | 0.22580645161290325 | 0.21774193548387105 | 0.057142857142857176 | 0.75 | 0.5| True | False | True | False | False | True | True | False | False | True | False | True | False | True | False | True | True | False | True | False | True | False |
| 0.743243 | 0.5 | 0.44206008583690987 | 0.9019607843137255 | 0.4320987654320988 | 0.5 | 0.7931034482758623 | 0.0 | 0.032258064516129004 | 0.39516129032258074 | 0.057142857142857176 | 0.5 | 0.0| True | False | True | False | True | False | True | False | False | True | False | True | False | True | False | True | True | False | True | False | True | False |
| 0.729730 | 0.75| 0.1502145922746781 | 0.28104575163398693 | 0.2345679012345679 | 0.5333333333333332 | 0.7931034482758623 | 0.3366336633663366 | 0.32258064516129026 | 0.5000000000000001 | 0.3142857142857142 | 0.5 | 0.0| True | False | False | True | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False |
| 0.540541 | 0.0 | 0.39914163090128757 | 0.5359477124183006 | 0.35802469135802467 | 0.7000000000000002 | 0.3793103448275863 | 0.207920792079208 | 0.16129032258064513 | 0.8306451612903228 | 0.057142857142857176 | 0.25 | 0.0| False | True | False | True | True | False | True | False | False | True | False | True | False | True | True | False | True | False | True | False | True | False |
| 0.675676 | 0.75| 0.25321888412017163 | 0.6339869281045752 | 0.7777777777777779 | 0.36666666666666625 | 0.6551724137931034 | 0.13861386138613863 | 0.19354838709677413 | 0.1693548387096774 | 0.11428571428571424 | 0.5 | 0.0| True | False | True | False | True | False | True | False | False | True | True | False | True | False | True | False | True | False | True | False | True | False |
| 0.716216 | 0.5 | 1.0 | 0.16339869281045755 | 0.11111111111111113 | 0.06666666666666643 | 0.2068965517241379 | 0.26732673267326745 | 0.38709677419354827 | 0.5322580645161291 | 0.3714285714285714 | 0.25 | 0.0| True | False | False | True | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False |
| 0.729730 | 0.0 | 0.9356223175965666 | 0.16993464052287582 | 0.16049382716049385 | 0.33333333333333304 | 0.034482758620689724 | 0.01980198019801982 | 0.06451612903225801 | 0.8790322580645163 | 0.0 | 0.75 | 0.25| False | True | True | False | False | True | True | False | False | True | True | False | True | False | True | True | False | True | False | True | True | False |
| 0.0 | 0.0 | 0.10300429184549353 | 0.37254901960784315 | 0.07407407407407407 | 0.5 | 0.6896551724137934 | 0.2178217821782179 | 0.22580645161290325 | 1.0 | 0.5142857142857142 | 1.0 | 0.0| True | False | True | False | True | False | False | True | True | False | True | False | True | False | True | True | False | True | False | True | True | False |
| 0.878378 | 0.0 | 0.20600858369098712 | 0.7516339869281046 | 0.6049382716049383 | 0.5333333333333332 | 0.6896551724137934 | 0.36633663366336633 | 0.38709677419354827 | 0.8790322580645163 | 0.3714285714285714 | 1.0 | 0.0| False | True | False | True | True | False | True | False | False | True | False | True | False | True | True | False | True | True | False | True | True | False |
| 0.851351 | 0.25| 0.6180257510729614 | 0.5620915032679739 | 0.7283950617283951 | 0.0 | 0.3448275862068966 | 0.16831683168316836 | 0.16129032258064513 | 0.5806451612903226 | 0.08571428571428565 | 1.0 | 0.75| False | True | True | False | False | True | False | True | False | True | False | True | False | True | False | True | False | True | False | True | True | False |
| 0.878378 | 0.25| 0.6394849785407726 | 0.47058823529411764 | 0.39506172839506176 | 0.43333333333333357 | 0.517241379310345 | 0.26732673267326745 | 0.322580645161290


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [74]:
# Code to rename all the columns in the dataset
import pandas as pd

# Sample DataFrame creation (you would normally load your dataset here)
# df = pd.read_csv('your_dataset.csv')

# Sample data
data = {
    'age': [0.202703, 0.905405],
    'bp': [0.75, 1.00],
    'sg': [1, 0],
    'al': [0, 0],
    'su': [0, 0],
    'rbc': [1, 0],
    'pc': [0, 0],
    'pcc': [0, 0],
    'ba': [0, 0],
    'bgr': [0.158798, 0.965665],
    'bu': [0.196078, 0.522876],
    'sc': [0.160494, 0.641975],
    'sod': [0.166667, 0.666667],
    'pot': [0.206897, 0.000000],
    'hemo': [0.059406, 0.148515],
    'pcv': [0.000000, 0.225806],
    'wc': [0.653226, 0.217742],
    'rc': [0.500000, 0.500000],
    'htn': [0, 1],
    'dm': [0, 0],
    'cad': [1, 0],
    'appet': [1, 0],
    'pe': [0, 1],
    'ane': [1, 0],
    'class': [1, 0]
}

# Create the DataFrame
dataset = pd.DataFrame(data)

# Dictionary for renaming the columns
rename_dict = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Step to rename the columns
dataset.rename(columns=rename_dict, inplace=True)

# Display the updated DataFrame
display(dataset.head())



,age,blood pressure,specific gravity,albumin,sugar,red blood cells,pus cell,pus cell clumps,bacteria,blood glucose random,...,packed cell volume,white blood cell count,red blood cell count,hypertension,diabetes mellitus,coronary artery disease,appetite,pedal edema,anemia,class
0,0.202703,0.75,1,0,0,1,0,0,0,0.158798,...,0.000000,0.653226,0.5,0,0,1,1,0,1,1
1,0.905405,1.00,0,0,0,0,0,0,0,0.965665,...,0.225806,0.217742,0.5,1,0,0,0,1,0,0


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [75]:
# Converted SQL to Pandas code
import pandas as pd

# Sample DataFrame setup
data = {'Target': ['ckd', 'notckd', 'ckd', 'notckd', 'ckd']}
dataset = pd.DataFrame(data)

# Pandas query to replicate the SQL query
result = dataset.groupby('Target').size().reset_index(name='count')

# Display the results
display(result)


,Target,count
0,ckd,3
1,notckd,2


*Describe what the above code does here*

The code above finds the target data by defining it as "data". It then groups it by the target and then categorizes it with the count and then displays the result. It counts how many times each unique target shows up in the dataset.